---
title: "Detecting and Correcting for Batch-Effects in scRNAseq Data"
author: "Nikolay Oskolkov"
date: "September 26, 2017"
output:
  html_document:
    toc: yes
  md_document:
    variant: markdown_github
abstract: |
  In this lab we will check for presence of batch-effects in the data set and identify genes significantly affected by batch-effects. Further we will show ways removal technical variation due to the batch-effects.
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir="/home/nikolay/Documents/Teaching/scRNAseq_course/")
```

### Load packages

First of all we will  R packages which we are going to use in this lab:

```{r, message=FALSE}
library("Seurat")
suppressMessages(require(gplots))
suppressMessages(require(ggplot2))
suppressMessages(require(matrixStats))
suppressMessages(require(mixOmics))
suppressMessages(require(RColorBrewer))
suppressMessages(require(Rtsne))
```

### Load Expression Values and Metadata

Let start with loading the matrix of raw gene expression counts and filtering away genes with median count across all cells below 1, it is very conservative but speeds up computations for this lab. Those genes are lowly expressed genes which should be excluded from the downstream analysis as they might lead to spurious results:

```{r loading and filtering expression data}
D <- read.table("data/ILC/ensembl_countvalues_ILC.csv",sep=",",header=T,row.names=1)
library("matrixStats")
D<-D[rowMedians(as.matrix(D))>=1,]
D[1:5,1:5]
dim(D)
```

For the sake of speed and simplicity of this lab we will select only 50% of most varying genes using coefficient of variation as a criterion:

```{r selecting most variable genes}
D_var<-apply(D,1,function(x) sd(x)/mean(x))
D<-D[D_var>quantile(D_var,0.5),]
D[1:5,1:5]
dim(D)
```

The rows of the matrix represent Ensembl gene IDs (you can convert them to gene symbols using biomaRt package) from `r dim(D)[1]` genes, and the columns are IDs from `r dim(D)[2]` cells from different individuals sequenced at different plates. To see how many individuals and plates we have let us load the meta-information and have a look:

```{r}
M <- read.table("data/ILC/Metadata_ILC.csv",sep=",",header=T,row.names=1)
M$Plate<-matrix(unlist(strsplit(as.character(M$Plate),"_")),byrow=TRUE,ncol=2)[,2]
head(M)
summary(M)
```

Thus we have cells from `r length(levels(factor(M$Donor)))` individuals with IDs:

```{r}
levels(factor(M$Donor))
```

that were pooled together and sequenced at `r length(levels(factor(M$Plate)))` plates with IDs:

```{r}
levels(factor(M$Plate))
```

and finally we have `r length(levels(factor(M$Celltype)))` cell-types with the following IDs:

```{r}
levels(factor(M$Celltype))
```

The variables Plate and Donor might potentially reflect technical variation due to batch. 


### Checking for Genome-Wide Batch-Effects

Now let us check potential batch-effects in the data set. As we saw previously the cells were pooled from `r length(levels(factor(M$Donor)))` and sequenced on `r length(levels(factor(M$Plate)))` plates. Thoese are potential batches. We need to check how they affect gene expression genome-wide. One way to see it is to plot PCA and tSNE and color cells by batch:

```{r PCA and tSNE,fig.width=10,fig.height=8}
library("mixOmics")
pca.ILC<-pca(log10(t(D+1)),ncomp=10,center=TRUE,scale=FALSE)
plot(pca.ILC)
plotIndiv(pca.ILC,group=factor(M$Plate),ind.names=FALSE,ellipse=FALSE,legend=TRUE,title="PCA PLOT, PLATE EFFECT",cex=1)
plotIndiv(pca.ILC,group=factor(M$Donor),ind.names=FALSE,ellipse=FALSE,legend=TRUE,title="PCA PLOT, DONOR EFFECT",cex=1)
library("Rtsne")
library("RColorBrewer")
set.seed(1)
tsne.out_expr<-Rtsne(t(log10(D+1)),initial_dims=20,verbose=TRUE,perplexity=30)
palette(brewer.pal(length(levels(factor(M$Plate))),'Dark2'))
plot(tsne.out_expr$Y,main="tSNE PLOT, PLATE EFFECT",col=factor(M$Plate),xlab="tSNE1",ylab="tSNE2")
legend("topleft",levels(factor(M$Plate)),cex=1,fill=brewer.pal(length(levels(factor(M$Plate))),'Dark2'),inset=0.02)
palette(brewer.pal(length(levels(factor(M$Donor))),'Dark2'))
plot(tsne.out_expr$Y,main="tSNE PLOT, DONOR EFFECT",col=factor(M$Donor),xlab="tSNE1",ylab="tSNE2")
legend("topleft",levels(factor(M$Donor)),cex=1,fill=brewer.pal(length(levels(factor(M$Donor))),'Dark2'),inset=0.02)
```

We can immediately see that there is a slight plate related and a more pronounced donor related batch-effect. To further quantify it let us display how much of variation in each principal component is explained by the batch variables:

```{r heatmap batch effects,fig.width=10,fig.height=8}
M$Plate<-factor(M$Plate)
M$Donor<-factor(M$Donor)
M$Celltype<-factor(M$Celltype)

pc_adj_r_squared<-matrix(NA,ncol=dim(pca.ILC$x)[2],nrow=dim(M)[2])
for(i in 1:dim(pca.ILC$x)[2])
{
  print(i)
  for(j in 1:dim(M)[2])
  {
    pc_adj_r_squared[j,i]<-summary(lm(pca.ILC$x[,i]~M[,j]))$adj.r.squared
  }
}
pc_adj_r_squared<-as.data.frame(pc_adj_r_squared)
colnames(pc_adj_r_squared)<-colnames(pca.ILC$x)
rownames(pc_adj_r_squared)<-colnames(M)
pc_adj_r_squared

library("gplots")
my_palette <- colorRampPalette(c("red", "yellow", "green"))(n = 299)
heatmap.2(data.matrix(pc_adj_r_squared),cellnote=round(pc_adj_r_squared,3),notecol="black",density.info="none",trace="none",col = my_palette, margins=c(8,10),dendrogram="row",Colv="NA",scale="row",main="ILC scRNAseq",cexRow=1,cexCol=1)
mtext("Adjusted R^2 of Association between PCs and Phenotypes")
```

From the heatmap above it is clear that 44% of PC1 is explained by Donor batch (Celltype contribution is negligible) while 31% of PC2 is explained by Plate batch. In addition, PC2 variation is mainly due to both Celltype and Plate batch, so those two variables are coupled and removing Plate batch could affect the variation due to Celltype which is our phenotype of interest.


### Checking for Per-Gene Batch-Effects

Now let us check batch-effects in the individual genes and figure out genes that are most influenced by batch. Let us check the effect of e.g. Donor on the expression of individual genes. For this purpose we will add a batch variable to the meta information:

```{r}
M$batch<-M$Donor
head(M)
```

Now we will rank all genes by the percentage of variation in their expression explained by the batch factor variable:

```{r genes affected by batch,fig.width=10,fig.height=8}
adj_r_squared<-vector()
for(i in 1:dim(D)[1])
{
  adj_r_squared<-append(adj_r_squared,summary(lm(as.numeric(D[i,])~M$batch))$adj.r.squared)
}
adj_r_squared[adj_r_squared<0]<-0
var_expl<-data.frame(genes=rownames(D),var_expl=adj_r_squared)
var_expl<-var_expl[order(-var_expl$var_expl),]
head(var_expl,20)
barplot(var_expl$var_expl[1:7],names=var_expl$genes[1:7],ylab="Variance Explained",main="Top Genes Influenced by Batch",col="darkred",las=1,cex.names=0.7,ylim=c(0,0.6))
```

Thus we conclude that the batch-effect might explain up to ~60% of variation in gene expression of most affected genes. 

Let us also check which batch is the most influential:

```{r most influential batch,fig.width=10,fig.height=8}
my_batches<-levels(M$batch)
my_genes<-as.character(var_expl$genes)
adj_r_squared_per_species<-list()
for(j in 1:length(my_genes))
{
  adj_r_squared_per_batch<-vector()
  for(i in 1:length(my_batches))
  {
    this_batch<-factor(ifelse(as.character(M$batch)==my_batches[i],my_batches[i],paste0("NOT_",my_batches[i])))
    adj_r_squared_per_batch<-append(adj_r_squared_per_batch,summary(lm(as.numeric(D[my_genes[j],])~this_batch))$adj.r.squared)
    adj_r_squared_per_batch[adj_r_squared_per_batch<0]<-0
  }
  adj_r_squared_per_species[[j]]<-adj_r_squared_per_batch
}
batch_matrix<-matrix(unlist(adj_r_squared_per_species),ncol=length(my_batches),byrow=TRUE)
batch_df<-as.data.frame(batch_matrix)
rownames(batch_df)<-my_genes
colnames(batch_df)<-my_batches
batch_df[1:10,1:3]

n <- length(my_batches)
library("RColorBrewer")
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
barplot(t(batch_matrix[1:7,]),beside=TRUE,ylab="Variance Explained",names.arg=my_genes[1:7],legend=my_batches,col=col_vector[1:length(my_batches)],cex.names=0.7,xlab="Genes",main="Batch Contribution to Genes")
```

Overall, it seems that different genes are affected by contributions from different individuals.

However, what would be variance explained by chance alone? To elucidate this we will perform a number of shuffling of expression vectors for each gene individually and calculate the shuffled variance explained, i.e. the variance explained by chance for each gene. Further, we will plot the noise zone as three standard deviations beyond the mean of shuffled variance explained. 

```{r noise zone calculation,fig.width=10,fig.height=8}
N_shuffle_ind<-100
ranked_genes<-as.character(var_expl$genes)
shuffle_stat_ind<-list()
for(i in 1:length(ranked_genes))
{
  adj_r_squared_shuffle_ind<-vector()
  for(j in 1:N_shuffle_ind)
  {
    gene_shuffle<-
      as.numeric(D[ranked_genes[i],][sample(1:dim(D)[2],dim(D)[2])])
    adj_r_squared_shuffle_ind<-
      append(adj_r_squared_shuffle_ind,summary(lm(gene_shuffle~M$batch))$adj.r.squared)
  }
  adj_r_squared_shuffle_ind[adj_r_squared_shuffle_ind<0]<-0
  shuffle_stat_ind[[i]]<-adj_r_squared_shuffle_ind
}
shuffle_matrix_ind<-t(matrix(unlist(shuffle_stat_ind),byrow=TRUE,ncol=N_shuffle_ind))
shuffle_matrix_ind[1:5,1:5]
library("matrixStats")
noise<-colMeans(shuffle_matrix_ind)+3*colSds(shuffle_matrix_ind)

library("ggplot2")
library("RColorBrewer")
Observed<-data.frame(ranked_genes=ranked_genes,var_expl=var_expl$var_expl)[1:6,]
ByChance<-data.frame(ranked_genes=ranked_genes,var_expl=noise)[1:6,]
ggplot(NULL, aes(ranked_genes,var_expl)) + 
  geom_bar(aes(fill="Observed"),data=Observed,stat='identity') +
  geom_bar(aes(fill="ByChance"),data=ByChance,stat='identity') +
  ggtitle("Observed vs. Resampled Variance Explained by Batch") +
  xlab("Rank of Genes") + ylab("Variance Explained") +
  scale_x_discrete(limits = ranked_genes[1:6], expand = c(0.01,0.01)) +
  scale_y_continuous(expand = c(0.0,0)) + coord_cartesian(ylim=c(0,0.6)) +
  theme(plot.title=element_text(hjust = 0.5)) + 
  scale_fill_manual(name="Legend",values=c(brewer.pal(8,'Dark2')[2],brewer.pal(8,'Dark2')[1]))
```

Above we displayed just a few most influenced by batch genes (observed) together with the shuffled variance explained (by chance). Here we can see that e.g. ENSG00000237973 seems to be strongly influenced by batch effects since the observed variance explained is beyond three standard deviations from the mean of variance explained by chance.

Now let us display the variance explained by batch for all genes in the ordered way (from largest to lowest) by a curve (observed) together with shuffled variance explained (by chance). Again, mean + 3 standard deviations from the mean is the noise zone boundary: 

```{r,fig.width=10,fig.height=8}
plot(var_expl$var_expl~seq(1:length(var_expl$var_expl)),xlab="Rank of Genes",ylab="Variance Explained",col="blue",main="Observed vs. Resampled Variance Explained by Batch",type="l",lwd=2,ylim=c(0,0.2))
lines(noise~seq(1:length(var_expl$var_expl)),col="red")
legend("topright",inset=0.02,c("Observed","ByChance"),col=c("blue","red"),lty=c(1,1))
```

Here we can see that the observed variance explained hits the noie zone for approximately gene #200 meaning that approximately top 200 genes ordered by their variance explained by batch are significantly influenced by batch, the rest genes are safe to use in the downstream analysis. We can also do a formal statistical test and calculate a p-value of significance of deviation from the noise zone. The p-value represents how many times shuffled variance explained by batch is equal or below the noise zone. We also apply Benjamini-Hochberg correction of th p-values for multiple testing:

```{r,fig.width=10,fig.height=8}
p_res<-vector()
for(i in 1:dim(shuffle_matrix_ind)[2])
{
  p_res<-append(p_res,sum(shuffle_matrix_ind[,i]>=var_expl$var_expl[i])/dim(shuffle_matrix_ind)[1])
}
p_res_BH<-p.adjust(p_res,method="BH")
plot(p_res_BH~seq(1:dim(shuffle_matrix_ind)[2]),type='l',col="darkgreen",xlab="Rank of Genes",ylab="p-value BH",main="Significance of Deviation from Noise")
abline(h=0.05,col="red")
```

Again, we see that the top ca. 200 genes seem to be significanly influenced by the batch-effects. Finally, let us display the genes that are significantly influenced by batch-effects to have a look and memorize them:

```{r}
problematic_genes<-data.frame(species=ranked_genes,var_expl_by_batch=var_expl$var_expl,pvalue=p_res,FDR=p_res_BH)
problematic_genes<-problematic_genes[order(problematic_genes$FDR,problematic_genes$pvalue,-problematic_genes$var_expl_by_batch),]
bad_genes<-problematic_genes[problematic_genes$FDR<=0.05,]
good_genes<-problematic_genes[problematic_genes$FDR>0.05,]
```

Thus here are "bad genes" ordered by how strongly they are affected by batch-effects, i.e. the higher in the list the more affected:

```{r}
head(bad_genes,50)
dim(bad_genes)[1]
```

And here come genes that are ok to use for the downstream analysis since they are not significantly affected by batch effects:

```{r}
head(good_genes,50)
dim(good_genes)[1]
```


### Correcting for Batch-Effects with Combat

Here we will use Combat in order to remove batch-effects due to different Donors. Combat has an advantage compared to simple linear regression batch-effects removal implemented e.g. in Limma since it uses Bayesian framework and works well even for low numbers of samples/cells. Particular advantage of the Combat batch effect removal is that it uses so-called “Bayesian shrinkage towards the mean”, i.e. correction of each gene uses a “mean” information from other genes which Limma does Frequentist batch-effect correction gene-by-gene independently. We assume that the different Donor and Plate can be viewed as a bacth which can be removed with Combat.

An important step to do before running Combat is to show the phenotype of interst (which is Celltype in our case) to the algorithm. In this way we are saying: do not touch the variation due to the Celltype, correct for the differences in Donor only. 

```{r Combat,fig.width=10,fig.height=8}
library("sva")
modcombat = model.matrix(~as.factor(as.character(M$Celltype)), data=M)
combat_edata = ComBat(dat=D, batch=M$batch, mod=modcombat, par.prior=TRUE, prior.plots=TRUE)
D[1:5,1:5]
combat_edata[1:5,1:5]
```

Let us now display a tSNE plot for the Combat corrected data set:

```{r tSNE Combat,fig.width=10,fig.height=8}
library("Rtsne")
set.seed(1)
tsne.out_combat<-Rtsne(t(combat_edata),initial_dims=10,verbose=TRUE,perplexity=30)

#BEFORE CORRECTION
plot(tsne.out_expr$Y,main="ILC, tSNE PLOT, BEFORE COMBAT",col=factor(M$Donor),xlab="tSNE1",ylab="tSNE2")
legend("topleft",levels(factor(M$Donor)),cex=1,fill=brewer.pal(length(levels(factor(M$Donor))),'Dark2'),inset=0.02)

#AFTER COMBAT CORRECTION
plot(tsne.out_combat$Y,main="ILC, tSNE PLOT, AFTER COMBAT",col=as.factor(M$Donor),xlab="tSNE1",ylab="tSNE2")
legend("topright",levels(as.factor(M$Donor)),cex=1,fill=unique(as.factor(M$Donor)),inset=0.02)
```

We see that cells from different Donors seem to be more mixed after Combat correction.


### Correcting for Batch-Effects with MNN

Now we will perform batch-effects correction with Mutial Nearest Neighbours (MNN) algorithm from the "scran" R package developed in the lab of John Marioni, University of Cambridge. The idea of the method is to identify cells between batches having shortest Euclidean distance and hence (presumably!) belonging to the same sub-population (nearest neighbours). Then the difference in expression between the cells is assumed due to technical/batch variation and all other cells are corrected with respect to the identified nearest neighbour cells.

For simplicity we assume that T74 and T75 belong to the first batch while T86 is the second batch.

```{r MNN,fig.width=10,fig.height=8}
library("scran")
D_T74<-subset(D,select=colnames(D)[grepl("T74",colnames(D))])
D_T75<-subset(D,select=colnames(D)[grepl("T75",colnames(D))])
D_T86<-subset(D,select=colnames(D)[grepl("T86",colnames(D))])
mnn_corrected <- mnnCorrect(as.matrix(D_T74), as.matrix(D_T75),as.matrix(D_T86))
mnn_corrected$corrected[[1]][1:5,1:5]

mnn_merge<-cbind(mnn_corrected$corrected[[1]],mnn_corrected$corrected[[2]],mnn_corrected$corrected[[3]])
dim(mnn_merge)
library("Rtsne")
set.seed(1)
tsne.out_mnn<-Rtsne(t(mnn_merge),initial_dims=10,verbose=TRUE,perplexity=30)
plot(tsne.out_mnn$Y,main="ILC, tSNE PLOT, MNN CORRECTION",col=factor(M$Donor),xlab="tSNE1",ylab="tSNE2")
legend("topleft",levels(factor(M$Donor)),cex=1,fill=brewer.pal(length(levels(factor(M$Donor))),'Dark2'),inset=0.02)
```

### Correcting for Batch-Effects with SCMAP

Now we are going to use a method called SCMAP for integration of two scRNAseq data sets which can be view as a correction for batch-effects. The idea of the method is to use one scRNAseq data set as a reference (e.g. a Human Cell Atlas data can be considered as a reference) and project cells from our particular scRNAseq experiment onto the reference data in order to determine the cell types in our scRNAseq data.

For simplicity we will consider cells from T74 and T75 donors as a reference and project cells from T86 onto the reference. Let us start with creating a SingleCellExperiment object for the reference data set.

```{r Create SCE object}
library("SingleCellExperiment")
library("scmap")
ref_df<-subset(D,select=colnames(D)[grepl("T74|T75",colnames(D))])
ref_annot<-M[grepl("T74|T75",as.character(M$Donor)),]
ref_annot_scmap<-data.frame(cell_type1=ref_annot$Celltype)
rownames(ref_annot_scmap)<-rownames(ref_annot)
head(ref_annot_scmap)  

sce <- SingleCellExperiment(assays = list(normcounts = as.matrix(ref_df)), colData = ref_annot_scmap)
logcounts(sce) <- log2(normcounts(sce) + 1)
# use gene names as feature symbols
rowData(sce)$feature_symbol <- rownames(sce)
isSpike(sce, "ERCC") <- grepl("^ERCC-", rownames(sce))
# remove features with duplicated names
sce <- sce[!duplicated(rownames(sce)), ]
sce
```

Now we will select most informative genes that provide cell clustering in the reference data set:

```{r Select Features,fig.width=10,fig.height=8}
sce <- selectFeatures(sce, suppress_plot = FALSE)
table(rowData(sce)$scmap_features)
```

Here we index reference data set clustering using medians of genes for each cluster:

```{r Heatmap,fig.width=10,fig.height=8}
sce <- indexCluster(sce)
head(metadata(sce)$scmap_cluster_index)
heatmap(as.matrix(metadata(sce)$scmap_cluster_index))
```

Lets us now create a SCE object for the test data set tp be projected onto the reference:

```{r Test DF}
test_df<-subset(D,select=colnames(D)[grepl("T86",colnames(D))])
sce_test <- SingleCellExperiment(assays = list(normcounts = as.matrix(test_df)))
logcounts(sce_test) <- log2(normcounts(sce_test) + 1)
# use gene names as feature symbols
rowData(sce_test)$feature_symbol <- rownames(sce_test)
isSpike(sce_test, "ERCC") <- grepl("^ERCC-", rownames(sce_test))
# remove features with duplicated names
sce_test <- sce_test[!duplicated(rownames(sce_test)), ]
sce_test
```

Now we can project the test data set onto the refernce:

```{r Celltype Assignment}
scmapCluster_results <- scmapCluster(
  projection = sce_test, threshold = 0.3,
  index_list = list(
    reference = metadata(sce)$scmap_cluster_index
  )
)
head(scmapCluster_results$scmap_cluster_labs,20)
head(scmapCluster_results$scmap_cluster_siml,20)
results<-data.frame(CELS=colnames(test_df),ASSIGNED_LABEL=as.vector(scmapCluster_results$scmap_cluster_labs),TRUE_LABEL=matrix(unlist(strsplit(colnames(test_df),"_")),ncol=4,byrow=TRUE)[,4],SIMILARITY=as.vector(scmapCluster_results$scmap_cluster_siml))
head(results,20)
```

We can see that there are `r sum(as.character(results$ASSIGNED_LABEL)=="unassigned")` "unassigned" cells, i.e. where KNN classifier fails to assign their class. Let as calculate the accuracy of assignment for the assigned cells:

```{r Accuracy}
results_assigned<-results[as.character(results$ASSIGNED_LABEL)!="unassigned",]
head(results_assigned,20)
table(results_assigned$ASSIGNED_LABEL,results_assigned$TRUE_LABEL)
sum(as.character(results_assigned$ASSIGNED_LABEL)==as.character(results_assigned$TRUE_LABEL))/dim(results_assigned)[1]
```

We conclude that the accuracy of assignment is 84% which is not fantastic taking into account that SCMAP failed assignment of almost a half of the cells in the test data set. 


### Correcting for Batch Effects with Seurat

Now we will use Seurat “alignment” method which is based on Canonical Correlation Analysis (CCA). CCA extract common sources of variation and facilitate identification of common cell types between two data sets.

```{r Seurat,fig.width=10,fig.height=8}
library("Seurat")
ref_df<-subset(D,select=colnames(D)[grepl("T74|T75",colnames(D))])
test_df<-subset(D,select=colnames(D)[grepl("T86",colnames(D))])

# Set up reference object
ref_seurat <- CreateSeuratObject(raw.data = ref_df, project = "ref", min.cells = 5)
ref_seurat@meta.data$stim <- "ref"
ref_seurat <- FilterCells(ref_seurat, subset.names = "nGene", low.thresholds = 100, high.thresholds = Inf)
ref_seurat <- NormalizeData(ref_seurat)
ref_seurat <- ScaleData(ref_seurat, display.progress = T)

# Set up T75 object
test_seurat <- CreateSeuratObject(raw.data = test_df, project = "test", min.cells = 5)
test_seurat@meta.data$stim <- "test"
test_seurat <- FilterCells(test_seurat, subset.names = "nGene", low.thresholds = 100, high.thresholds = Inf)
test_seurat <- NormalizeData(test_seurat)
test_seurat <- ScaleData(test_seurat, display.progress = T)

# Gene selection for input to CCA
ref_seurat <- FindVariableGenes(ref_seurat, do.plot = F)
test_seurat <- FindVariableGenes(test_seurat, do.plot = F)
g.1 <- head(rownames(ref_seurat@hvg.info), 1000)
g.2 <- head(rownames(test_seurat@hvg.info), 1000)
genes.use <- unique(c(g.1, g.2))
genes.use <- intersect(genes.use, rownames(ref_seurat@scale.data))
genes.use <- intersect(genes.use, rownames(test_seurat@scale.data))

#CCA
seurat_combined <- RunCCA(ref_seurat, test_seurat , genes.use = genes.use, num.cc = 30)
p1 <- DimPlot(object = seurat_combined,reduction.use="cca", group.by = "stim", pt.size=0.5, do.return=TRUE)
p2 <- VlnPlot(object = seurat_combined, features.plot = "CC1", group.by = "stim", do.return = TRUE)
plot_grid(p1,p2)
PrintDim(object = seurat_combined, reduction.type = "cca", dims.print = 1:2, genes.print = 10)
p3 <- MetageneBicorPlot(seurat_combined, grouping.var = "stim", dims.eval = 1:30, display.progress = FALSE)
DimHeatmap(object = seurat_combined, reduction.type = "cca", cells.use = 500, dim.use = 1:9, do.balanced = TRUE)
seurat_combined <- AlignSubspace(seurat_combined, reduction.type = "cca", grouping.var = "stim", dims.align = 1:20)

# t-SNE and Clustering
seurat_combined <- RunTSNE(seurat_combined, reduction.use = "cca.aligned", dims.use = 1:20, do.fast = T)
seurat_combined <- FindClusters(seurat_combined, reduction.type = "cca.aligned", resolution = 0.6, dims.use = 1:20,print.output = FALSE)

# Visualization
p1 <- TSNEPlot(seurat_combined, do.return = T, pt.size = 0.5, group.by = "stim")
p2 <- TSNEPlot(seurat_combined, do.label = T, do.return = T, pt.size = 0.5)
plot_grid(p1, p2)
```

Again we see good mixing of cells from the refernce and test data sets.

### Session Information

Finally here is the details on the system on which this document was compiled:

```{r}
sessionInfo()
```
